raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -580,9 +585,15 @@ private EndQuorumEpochResponseData handleEndQuorumEpochRequest(
             if (state.isUnattached()
                 || state.hasLeader(requestReplicaId)
                 || state.hasVotedFor(requestReplicaId)) {
-                // reset the timer to election jitter so that they would become candidate
-                // after a pretty short period of time
-                timer.reset(randomElectionJitterMs());
+                List<Integer> preferredSuccessors = request.preferredSuccessors();
+                // We didn't find the corresponding voters inside the request.
+                if (!preferredSuccessors.contains(quorum.localId)) {
+                    return buildEndQuorumEpochResponse(Errors.INCONSISTENT_VOTER_SET);
+                }
+                // Based on the priority inside the preferred successors, choose the corresponding delayed
+                // election time so that the most up-to-date voter has a higher chance to be elected.
+                // The reasoning for always making it positive is for deterministic unit testing.
+                timer.reset(electionJitterMs * preferredSuccessors.indexOf(quorum.localId) + positiveRandomElectionJitterMs());
```

Comment:
The preferred successor should start the election immediately I think? It seems like this logic will always add some jitter. On a more general level, I wonder how much value there is being able to provide a list of potential successors vs a single preferred one. I am not sure how this will be implemented with Guozhang's revised timeout semantics.

cc @guozhangwang

Reply:
Having the entire successor list helps better on the strict ordering based on the offsets in this case I think, compared with only getting one node start the election while others still have a chance to start election concurrently.

Reply:
I'm mostly just concerned about the awkward backoff logic with multiple successors. Seems like it is simpler to understand if we have just one preferred successor who becomes candidate immediately and everyone else waits the normal jitter time. In general, EndQuorumEpoch is trying to handle the common graceful shutdown case. Not sure it's worth trying to optimize for the case when the preferred successor fails at the same time the leader wants to shutdown.

Reply:
I'm planning to add the bounded exponential backoff time for the case when election has failed, whereas the election timeout itself would be set the election timeout as uniform(T, 2T) where T is the configured value and hence remove the jitter config. Also the fetch timeout on the follower would also be set as uniform(T', 2T') where T' is the configured value. With that:

* When fetch timeout expired, a follower would become a candidate immediately and start sending vote with the new epoch.
* When an election failed but there's no learned leaders, the candidate would backoff, and then bump the epoch and resend vote again.
* [For this PR] When received the end epoch, the preferred one could try to become the candidate immediately while others still wait for the backoff.

Assuming we use a sorted list of preferred successor based on their log end offset / epoch, I think it is fine to let the Nth preferred successor to set its backoff as `(N -1)^2 * base-backoff`.